toolchain/base/value_store.h

```diff
@@ -221,62 +221,67 @@ auto CanonicalValueStore<IdT>::Reserve(size_t size) -> void {
   values_.Reserve(size);
 }
 
-// A ValueStore keyed on `FromIdT`, and maps that to a `StoredIdT`.
-template <typename FromIdT, typename StoredIdT>
+// A ValueStore that builds a 1:1 relationship between two IDs.
+// * `RelatedIdT` represents a related ID that can be used to find values in the
+//   store.
+// * `IdT` is the actual ID of values in this store, and `IdT::ValueType` is the
+//   value type being stored.
+//
+// The value store builds a mapping so that either ID can be used later to find
+// a value. And the user can query if a related `RelatedIdT` has been used to
+// add a value to the store or not.
+//
+// When adding to the store, the user provides the related `RelatedIdT` along
+// with the value being stored, and gets back the ID of the value in the store.
+template <typename RelatedIdT, typename IdT>
 class RelationalValueStore {
-  struct ConditionalId {
-    using ValueType = std::optional<typename StoredIdT::ValueType>;
-    StoredIdT id;
-  };
-
  public:
-  using RefType = StoredIdT::ValueType&;
-  using ConstRefType = const StoredIdT::ValueType&;
-
-  // Given the original ID and a value, stores the value and returns a mapped ID
-  // to reference it. This allows lookups later through either id.
-  auto Add(FromIdT from_id, StoredIdT::ValueType value) -> StoredIdT {
-    CARBON_DCHECK(from_id.index >= 0, "{0}", from_id);
-    StoredIdT id(from_id.index);
+  using ValueType = IdT::ValueType;
+  using RefType = ValueStore<IdT>::RefType;
+  using ConstRefType = ValueStore<IdT>::ConstRefType;
+
+  // Given the related ID and a value, stores the value and returns a mapped ID
+  // to reference it in the store.
+  auto Add(RelatedIdT related_id, ValueType value) -> IdT {
+    CARBON_DCHECK(related_id.index >= 0, "{0}", related_id);
+    IdT id(related_id.index);
     if (static_cast<size_t>(id.index) >= values_.size()) {
       values_.resize(id.index + 1);
     }
     values_[id.index] = std::move(value);
```

Comment:
Validate that this entry wasn't already added?

Reply:
Oh good idea, yes, done